README.md
# liphdbpp-mysql

[![TangoControls](https://img.shields.io/badge/-Tango--Controls-7ABB45.svg?style=flat&logo=%20data%3Aimage%2Fpng%3Bbase64%2CiVBORw0KGgoAAAANSUhEUgAAACAAAAAkCAYAAADo6zjiAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEwAACxMBAJqcGAAAAsFJREFUWIXtl01IFVEYht9zU%2FvTqOxShLowlOgHykWUGEjUKqiocB1FQURB0KJaRdGiaFM7gzZRLWpTq2olhNQyCtpYCP1gNyIoUTFNnxZzRs8dzvw4Q6564XLnfOf73vedc2a%2BmZEKALgHrC3CUUR8CxZFeEoFalsdM4uLmMgFoIlZLJp3A9ZE4S2oKehhlaR1BTnyg2ocnW%2FxsxEDhbYij4EPVncaeASMAavnS%2FwA8NMaqACNQCew3f4as3KZOYh2SuqTVJeQNiFpn6QGSRVjTH9W%2FiThvcCn6H6n4BvQDvQWFT%2BSIDIFDAKfE3KOAQeBfB0XGPeQvgE67P8ZoB44DvTHmFgJdOQRv%2BUjc%2BavA9siNTWemgfA3TwGquCZ3w8szFIL1ALngIZorndvgJOR0GlP2gtJkzH%2Bd0fGFxW07NqY%2FCrx5QRXcYjbCbmxF1dkBSbi8kpACah3Yi2Sys74cVyxMWY6bk5BTwgRe%2BYlSzLmxNpU3aBeJogk4XWWpJKUeiap3RJYCpQj4QWZDQCuyIAk19Auj%2BAFYGZZjTGjksaBESB8P9iaxUBIaJzjZcCQcwHdj%2BS2Al0xPOeBYYKHk4vfmQ3Y8YkIwRUb7wQGU7j2ePrA1URx93ayd8UpD8klyPbSQfCOMIO05MbI%2BDvwBbjsMdGTwlX21AAMZzEerkaI9zFkP4AeYCPBg6gNuEb6I%2FthFgN1KSQupqzoRELOSed4DGiJala1UmOMr2U%2Bl%2FTWEy9Japa%2Fy41IWi%2FJ3d4%2FkkaAw0Bz3AocArqApwTvet3O3GbgV8qqjAM7bf4N4KMztwTodcYVyelywKSCD5V3xphNXoezuTskNSl4bgxJ6jPGVJJqbN0aSV%2Bd0M0aO7FCs19Jo2lExphXaTkxdRVgQFK7DZVDZ8%2BcpdmQh3wuILh7ut3AEyt%2B51%2BL%2F0cUfwFOX0t0StltmQAAAABJRU5ErkJggg%3D%3D)](http://www.tango-controls.org) [![License: LGPL v3](https://img.shields.io/badge/License-LGPL%20v3-blue.svg)](https://www.gnu.org/licenses/lgpl-3.0) [![GitHub release](https://img.shields.io/github/release/tango-controls-hdbpp/libhdbpp-mysql.svg)](https://github.com/tango-controls-hdbpp/libhdbpp-mysql/releases)   [![Download](https://api.bintray.com/packages/tango-controls/debian/libhdb%2B%2Bmysql6/images/download.svg)](https://bintray.com/tango-controls/debian/libhdb%2B%2Bmysql6/_latestVersion)


- [liphdbpp-mysql](#liphdbpp-mysql)
  - [Version](#Version)
  - [Documentation](#Documentation)
  - [Building](#Building)
    - [Dependencies](#Dependencies)
    - [Build](#Build)
    - [Build Flags](#Build-Flags)
  - [Installation](#Installation)
  - [DB Schema](#DB-Schema)
    - [Partition management](#Partition-management)
    - [MyIsam to InnoDB schema migration](#MyIsam-to-InnoDB-schema-migration)
    - [JSON arrays schema migration](#JSON-arrays-schema-migration)
    - [DevEnum: archiving of DevEnum labels](#DevEnum:-archiving-of-DevEnum-labels)
    - [TTL: per attribute old data deletion](#TTL:-per-attribute-old-data-deletion)
  - [Configuration](#Configuration)
    - [Library Configuration Parameters](#Library-Configuration-Parameters)
    - [Configuration Example](#Configuration-Example)
  - [License](#License)



Library for HDB++ implementing MySQL schema

## Version

The current release version is 7.1.0

## Documentation

* See the Tango documentation [here](http://tango-controls.readthedocs.io/en/latest/administration/services/hdbpp/index.html#hdb-an-archiving-historian-service) for broader information about the HDB++ archiving system and its integration into Tango Controls
* libhdbpp-mysql [CHANGELOG.md](https://github.com/tango-controls-hdbpp/libhdbpp/blob/master/CHANGELOG.md) contains the latest changes both released and in development.

## Building

### Dependencies

Ensure the development version of the dependencies are installed. These are as follows:

* HDB++ library libhdbpp (downloaded automatically by cmake if not found)
* Tango Controls 9 or higher - either via debian package or source install.
* omniORB release 4 - libomniorb4 and libomnithread.
* libzmq - libzmq3-dev or libzmq5-dev.
* MySQL client library - libmysqlclient-dev or libmariadb-dev (>=3)

### Build

The build system uses pkg-config to find some dependencies, for example Tango. If Tango is not installed to a standard location, set PKG_CONFIG_PATH, i.e.

```bash
export PKG_CONFIG_PATH=/non/standard/tango/install/location
```

Then to build the library:

```bash
mkdir -p build
cd build
cmake ..
make
```

If HDB++ library libhdbpp is not installed in system paths, CMAKE_PREFIX_PATH can be used to point to non standard install location, or CMAKE_INCLUDE_PATH should be set to point to the sources location. If neither of the previous condition is true, cmake will automatically fetch the dependency from the github repository.
CMake flags can be set on the command line at configuration time, i.e.:

```bash
cmake -DCMAKE_INCLUDE_PATH=/path/to/local/install/of/libhdbpp/headers .. 
```

The pkg-config path can also be set with the cmake argument CMAKE_PREFIX_PATH. This can be set on the command line at configuration time, i.e.:

```bash
cmake -DCMAKE_PREFIX_PATH=/non/standard/dependency/install/location .. 
```

### Build Flags

The following is a list of common useful CMake flags and their use:

| Flag | Setting | Description |
|------|-----|-----|
| CMAKE_INSTALL_PREFIX | PATH | Standard CMake flag to modify the install prefix. |
| CMAKE_PREFIX_PATH | PATH[S] | Standard CMake flag to add paths where to look for installed dependencies. |
| CMAKE_INCLUDE_PATH | PATH[S] | Standard CMake flag to add include paths to the search path. |
| CMAKE_LIBRARY_PATH | PATH[S] | Standard CMake flag to add paths to the library search path |
| CMAKE_BUILD_TYPE | Debug/Release | Build type to produce |


## Installation

Once built simply run `make install`. The install can be passed a PREFIX variable, this is set to /usr/local by default. It is also possible to use DESTDIR. Install path is constructed as DESTDIR/PREFIX.

## DB Schema

Three files are provided to create the DB schema:

* etc/create_hdb++_mysql.sql
* etc/create_hdb++_mysql_innodb.sql
* etc/create_hdb++_mysql_innodb_json.sql

In the first file MyISAM is used as the DB engine and 3 TIMESTAMP columns (data_time, recv_time, insert_time) are created for every data table.
MyIsam is deprecated and partitioning on MyIsam tables is no longer supported since Mysql 8.
In the second file InnoDB is used as the DB engine and 2 DATETIME columns (data_time, recv_time) are created for every data table.
A primary key (att_conf_id, data_time) is created for every data table, with this key maximum performances are obtained by partitioning tables by range on the data_time column.
At the same time with this primary key duplicated of data_time with the same att_conf_id are not allowed. To ignore primary key duplication errors, the 'ignore_duplicates' configuration key can be set to 1.
The third file, as the second one, uses InnoDB as the DB engine, 2 DATETIME columns and partitioning. But for array attributes, instead of storing every element of the array in a single row, stores all the elements of the array in a JSON field.
The use of JSON arrays needs to be enabled with json_array=1 in the LibConfiguration device/class property.
DB Schema, user creation script and partitioning procedure can be found in [hdbpp-mysql-project](https://github.com/tango-controls-hdbpp/hdbpp-mysql-project) under resources/schema.

### Partition management

Partitions defined in etc/create_hdb++_mysql_innodb.sql should be adjusted depending on the amount of data really present in the tables. Partitions 'p000' and 'future' should always be present to catch events arriving with the wrong (in the past or in the future) timestamps.
Older partitions can be dropped with the 'DROP PARTITION' command or moved to another database with the 'ALTER TABLE ... EXCHANGE PARTITION' command.
When approaching the end of the range of the current partition, the 'future' partition should be split into two parts to add the new partition with the 'ALTER TABLE ... REORGANIZE PARTITION' command.

### MyIsam to InnoDB schema migration

To migrate from the schema using MyIsam to the schema with InnoDB and partitiong, the suggested sequence is the following:

* rename existing tables (e.g. 'RENAME TABLE att_scalar_devboolean_ro TO att_scalar_devboolean_ro_old;')
* dump to file the renamed tables (e.g. 'SELECT ad.att_conf_id,ad.data_time,ad.recv_time, ad.value_r, ad.quality, ad.att_error_desc_id INTO OUTFILE '/tmp/att_scalar_devboolean_ro.csv' FROM hdbpp.att_scalar_devboolean_ro ad;')
* optimize MySQL configuration for InnoDB adjusting key_buffer_size, innodb_buffer_pool_size and many more parameters
* create new schema with 'etc/create_hdb++_mysql_innodb.sql'
* bulk load data from files into created tables (e.g. 'LOAD DATA LOCAL INFILE '/tmp/att_scalar_devboolean_ro.csv' INTO TABLE att_scalar_devboolean_ro;'). If the table size is big (in the order of tens of million rows) then the file to be loaded should be split into chunks of around 1 million rows: this can be easily done with scripts like 'pt-fifo-split' from Percona Tools.
* if everything went OK, old renamed tables can be dropped (e.g. 'DROP TABLE att_scalar_devboolean_ro_old;')

### JSON arrays schema migration

To migrate existing data from the schema using InnoDB and partitiong but multiple rows to store arrays elements to the schema using JSON for arrays, the suggested sequence is the following:

* rename existing array tables (e.g. 'RENAME TABLE att_array_devboolean_ro TO att_array_devboolean_ro_bk;')
* create the new array tables using the definitions in etc/create_hdb++_mysql_innodb_json.sql
* migrate using the appropriate queries (e.g. 'INSERT IGNORE INTO att_array_devboolean_ro (att_conf_id,data_time,recv_time,dim_x_r, dim_y_r,value_r,quality,att_error_desc_id) SELECT att_conf_id, data_time, recv_time, dim_x_r, dim_y_r, CONCAT('[',GROUP_CONCAT(value_r SEPARATOR ','),']'), quality, att_error_desc_id from att_array_devboolean_ro_bk GROUP BY att_conf_id,data_time;')

The file etc/hdb++_mysql_migrate_array_json.sql can be used for these operations.

### DevEnum: archiving of DevEnum labels

Support for DevEnum labels was added in a second time. If the column enum_labels in the table att_parameter is not present, it can be added executing the sql in the file etc/update_devenum_hdb++_mysql_innodb_json.sql

### TTL: per attribute old data deletion

In HDB++ a TimeToLive value in hours can be specified for each attribute. It is meant to let data older than the TTL value specified to be deleted. MySQL does not support this feature in a native way so something ad hoc needs to be implemented and run periodically.

The file etc/create_hdb++_mysql_delete_attr_ttl_procedure.sql can be used to create two stored procedure:
 * delete_attr_ttl which takes as parameters the attribute name and the ttl value
 * delete_ttl which takes no parameters and call delete_attr_ttl for all the attributes which ttl is greater than 0
 
 Note that it is preferable to use partitioning and drop the whole old partition. In fact deleting single values in MySQL can have a bad inpact on performances for the following reasons:

* deleting values in MySQL is always a heavy operation
* if MyISAM is used as DB engine, holes are left in tables after deletion. Tables with holes needs to be optimized (which can take quite a long time), otherwise performances degrade and an exclusive lock on the whole table needs to be taken for every operation (also for inserts).

## Configuration

This is the configuration to use for [event subscribers](https://github.com/tango-controls-hdbpp/hdbpp-es) to use this backend

### Library Configuration Parameters

Configuration parameters are as follows:

| Parameter | Mandatory | Default | Description |
|------|-----|-----|-----|
| libname | true | None | Must be "libhdb++mysql.so", a specific version or the full path can be specified |
| host | true | None | MySQL/MariaDB hostname or IP address |
| user | true | None | MySQL/MariaDB username |
| password | true | None | MySQL/MariaDB password |
| dbname | true | None | MySQL/MariaDB db name (e.g. hdbpp) |
| port | true | None | MySQL/MariaDB port (e.g. 3306) |
| lightschema | false | 0 | When set to 1 (true) insert_time and rcv_time are not stored. When set to 0 (false) presence of insert_time, rcv_time, quality, error_desc_id columns are auto-detected at library startup per table |
| ignore_duplicates | false | 0 | Schema optimized for InnoDB has the primary key on (att_conf_id,data_time), set this property to 1 to not have duplicates on (att_conf_id,data_time) signalled as storing errors. |
| json_array | false | 0 | When set to 1 use a JSON array to store Spectrum/Image data types instead of one row per array element. |
| batch_size | false | 1000 | Specify maximum batch size for insertion. |


### Configuration Example

Short example LibConfiguration property value on an EventSubscriber. You will HAVE to change the various parts to match your system:

```
host=mysql_hostname
user=hdbpprw
password=hdbpprw_pwd
dbname=hdbpp
port=3306
ignore_duplicates=1
json_array=1
batch_size=10000
libname=libhdb++mysql.so
```` 

## License

The code is released under the LGPL3 license and a copy of this license is provided with the code. 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [7.1.0]

### Added

* Support for DevEnum labels, needs update in att_parameter table.

## [7.0.0]

### Added

* Support for batch insertion.

### Changed

* Modified build system to use fetch libhdbpp and include it when requested.
* Made compatible with new libhdbpp (namespace, function and path changes)
* Code refactoring
* Align release with library version

## [1.2.1] - 2020-08-10

### Added

* Build using CMake.

### Changed

* Fixed bug in prepared statements after reconnection to the DB.
* Updated etc/hdb++_mysql_migrate_array_json.sql
* Updated README.md.

## [1.2.0] - 2018-04-04

### Added

* etc/create_hdb++_mysql_innodb_json.sql file to create the MySQL schema using InnoDB engine and time-range partitioning and JSON field for arrays.
* etc/hdb++_mysql_migrate_array_json.sql to migrate data from the tables with multiple rows per array element to the tables with JSON fields

### Changed

* LibHdb++MySQL.cpp: added json_array configuration key to store array elements in a JSON field of a single row
* Updated README.md.

## [1.1.0] - 2018-03-02

### Added

* etc/create_hdb++_mysql_innodb.sql file to create the MySQL schema using InnoDB engine and time-range partitioning.

### Changed

* LibHdb++MySQL.*: added ignore_duplicates configuration key to ignore duplicated key (att_conf_id,data_time) when inserting data in InnoDB schema
* Updated README.md.

## [1.0.0] - 2017-09-28

### Added

* CHANGELOG.md file.
* Debian Package build files under debian/

### Changed

* Makefile: Added install rules, clean up paths etc
* libhdb++ include paths to match new install location.
* Updated README.md.

### Removed

* libhdbpp submodule. This now has to be installed manually.


src/LibHdb++MySQL.h
//=============================================================================
//
// file :        LibHdb++MySQL.h
//
// description : Include for the LibHdb++MySQL library.
//
// Author: Graziano Scalamera
//
// $Revision: 1.1 $
//
// $Log: LibHdb++MySQL.h,v $
// Revision 1.1  2014-03-07 13:17:22  graziano
// first commit
//
//
//
//
//=============================================================================

#ifndef _HDBPP_MYSQL_H
#define _HDBPP_MYSQL_H

#include <mysql.h>
#include <hdb++/AbstractDB.h>

#include <string>
#include <iostream>
#include <sstream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <stdint.h>

//Tango:
#include <tango.h>
//#include <event.h>

#define TYPE_SCALAR					"scalar"
#define TYPE_ARRAY					"array"

#define TYPE_DEV_BOOLEAN			"devboolean"
#define TYPE_DEV_UCHAR				"devuchar"
#define TYPE_DEV_SHORT				"devshort"
#define TYPE_DEV_USHORT				"devushort"
#define TYPE_DEV_LONG				"devlong"
#define TYPE_DEV_ULONG				"devulong"
#define TYPE_DEV_LONG64				"devlong64"
#define TYPE_DEV_ULONG64			"devulong64"
#define TYPE_DEV_FLOAT				"devfloat"
#define TYPE_DEV_DOUBLE				"devdouble"
#define TYPE_DEV_STRING				"devstring"
#define TYPE_DEV_STATE				"devstate"
#define TYPE_DEV_ENCODED			"devencoded"
#define TYPE_DEV_ENUM				"devenum"

#define TYPE_RO						"ro"
#define TYPE_RW						"rw"

#define EVENT_ADD					"add"
#define EVENT_REMOVE				"remove"
#define EVENT_START					"start"
#define EVENT_STOP					"stop"
#define EVENT_CRASH					"crash"
#define EVENT_PAUSE					"pause"

//######## att_conf ########
#define CONF_TABLE_NAME				"att_conf"
#define CONF_COL_ID					"att_conf_id"
#define CONF_COL_NAME				"att_name"
#define CONF_COL_TYPE_ID			"att_conf_data_type_id"
#define CONF_COL_TTL				"att_ttl"
#define CONF_COL_FACILITY			"facility"
#define CONF_COL_DOMAIN				"domain"
#define CONF_COL_FAMILY				"family"
#define CONF_COL_MEMBER				"member"
#define CONF_COL_LAST_NAME			"name"

//######## att_conf_data_type ########
#define CONF_TYPE_TABLE_NAME		"att_conf_data_type"
#define CONF_TYPE_COL_TYPE_ID		"att_conf_data_type_id"
#define CONF_TYPE_COL_TYPE			"data_type"

//######## att_history ########
#define HISTORY_TABLE_NAME			"att_history"
#define HISTORY_COL_ID				"att_conf_id"
#define HISTORY_COL_EVENT_ID		"att_history_event_id"
#define HISTORY_COL_TIME			"time"

//######## att_history_event ########
#define HISTORY_EVENT_TABLE_NAME	"att_history_event"
#define HISTORY_EVENT_COL_EVENT_ID	"att_history_event_id"
#define HISTORY_EVENT_COL_EVENT		"event"

//######## att_scalar_... ########
#define SC_COL_ID					"att_conf_id"
#define SC_COL_INS_TIME				"insert_time"
#define SC_COL_RCV_TIME				"recv_time"
#define SC_COL_EV_TIME				"data_time"
#define SC_COL_VALUE_R				"value_r"
#define SC_COL_VALUE_W				"value_w"
#define SC_COL_QUALITY				"quality"
#define SC_COL_ERROR_DESC_ID		"att_error_desc_id"



//######## att_array_... ########
#define ARR_COL_ID					SC_COL_ID
#define ARR_COL_INS_TIME			SC_COL_INS_TIME
#define ARR_COL_RCV_TIME			SC_COL_RCV_TIME
#define ARR_COL_EV_TIME				SC_COL_EV_TIME
#define ARR_COL_VALUE_R				SC_COL_VALUE_R
#define ARR_COL_VALUE_W				SC_COL_VALUE_W
#define ARR_COL_IDX					"idx"
#define ARR_COL_DIMX_R				"dim_x_r"
#define ARR_COL_DIMY_R				"dim_y_r"
#define ARR_COL_DIMX_W				"dim_x_w"
#define ARR_COL_DIMY_W				"dim_y_w"
#define ARR_COL_QUALITY				SC_COL_QUALITY
#define ARR_COL_ERROR_DESC_ID		SC_COL_ERROR_DESC_ID

//######## att_error_desc ########
#define ERR_TABLE_NAME				"att_error_desc"
#define ERR_COL_ID					"att_error_desc_id"
#define ERR_COL_ERROR_DESC			"error_desc"


//######## att_parameter ########
#define PARAM_TABLE_NAME				"att_parameter"
#define PARAM_COL_ID					"att_conf_id"
#define PARAM_COL_INS_TIME				"insert_time"
#define PARAM_COL_EV_TIME				"recv_time"
#define PARAM_COL_LABEL					"label"
#define PARAM_COL_UNIT					"unit"
#define PARAM_COL_STANDARDUNIT			"standard_unit"
#define PARAM_COL_DISPLAYUNIT			"display_unit"
#define PARAM_COL_FORMAT				"format"
#define PARAM_COL_ARCHIVERELCHANGE		"archive_rel_change"
#define PARAM_COL_ARCHIVEABSCHANGE		"archive_abs_change"
#define PARAM_COL_ARCHIVEPERIOD			"archive_period"
#define PARAM_COL_DESCRIPTION			"description"
#define PARAM_COL_ENUM_LABELS			"enum_labels"

//######## INFORMATION SCHEMA ########
#define INFORMATION_SCHEMA				"INFORMATION_SCHEMA"
#define INF_SCHEMA_COLUMN_NAME			"COLUMN_NAME"
#define INF_SCHEMA_COLUMNS				"COLUMNS"
#define INF_SCHEMA_TABLE_SCHEMA			"TABLE_SCHEMA"
#define INF_SCHEMA_TABLE_NAME			"TABLE_NAME"

#define DEFAULT_BATCH_SIZE						1000 //TODO: configurable
#define MAX_PREPARED_STATEMENT_COUNT	16382 //to be less or equal to max-prepared-stmt-count configured on MySQL


namespace hdbpp
{
struct event_param
{
	string attr_name;
	int quality;/*ATTR_VALID, ATTR_INVALID, ..*/
	string error_desc;
	Tango::AttributeDimension attr_r_dim;
	Tango::AttributeDimension attr_w_dim;
	double ev_time;
	double rcv_time;
	bool isNull=false;
};

struct event_data_param
{
	Tango::EventData * data;
	event_param param;
};

template <typename T>
struct event_values_param
{
	vector<T> val_r;
	vector<T> val_w;
	event_param param;
};

class HdbPPMySQL : public AbstractDB
{
private:

	MYSQL *dbp;
	unsigned long db_mti;
	string m_dbname;
	unordered_map<string,int> attr_ID_map;
	bool lightschema;	//without recv_time and insert_time
	bool jsonarray;
	bool autodetectschema;
	bool ignoreduplicates;	//ignore duplicated key (att_conf_id,data_time) insert failures
	unsigned int batch_size{DEFAULT_BATCH_SIZE};
	unordered_map<string,int> attr_ERR_ID_map;
	queue<string> attr_ERR_queue;
	unordered_map<string,MYSQL_STMT *> pstmt_map;
	
	vector<Tango::CmdArgType> v_type;/*DEV_DOUBLE, DEV_STRING, ..*/
	vector<Tango::AttrDataFormat> v_format;/*SCALAR, SPECTRUM, ..*/
	vector<Tango::AttrWriteType> v_write_type;/*READ, READ_WRITE, ..*/
	unordered_map<string,bool > table_column_map;

	string get_data_type(int type/*DEV_DOUBLE, DEV_STRING, ..*/, int format/*SCALAR, SPECTRUM, ..*/, int write_type/*READ, READ_WRITE, ..*/);
	string get_table_name(int type/*DEV_DOUBLE, DEV_STRING, ..*/, int format/*SCALAR, SPECTRUM, ..*/, int write_type/*READ, READ_WRITE, ..*/);
	bool autodetect_column(string table_name, string column_name);
	int cache_ID(const string &attr_name, const string &func_name);
	bool cache_pstmt(const string &query, MYSQL_STMT **pstmt, unsigned int stmt_size, const string &func_name);
	void stmt_close(bool cached, const string &func, MYSQL_STMT	*pstmt);
	string err_stmt_close(bool cached, const string &func, const string &query, MYSQL_STMT *pstmt);
public:

	~HdbPPMySQL();

	HdbPPMySQL(const string &id,  const vector<string> &configuration);

	//void connect_db(string host, string user, string password, string dbname);
	int find_attr_id(string facility, string attr_name, int &ID);
	int find_attr_id_type(string facility, string attr_name, int &ID, string attr_type, unsigned int &conf_ttl);
	int find_last_event(int ID, string &event);
	int find_err_id(string err, int &ERR_ID);
	void cache_err_id(string error_desc, int &ERR_ID);
	int insert_error(string err, int &ERR_ID);
	// Inserts an attribute archive event for the EventData into the database. If the attribute
	// does not exist in the database, then an exception will be raised. If the attr_value
	// field of the data parameter if empty, then the attribute is in an error state
	// and the error message will be archived.
	void insert_event(Tango::EventData *event_data, const HdbEventDataType &data_type) override;

	// Insert multiple attribute archive events. Any attributes that do not exist will
	// cause an exception. On failure the fall back is to insert events individually
	void insert_events(std::vector<std::tuple<Tango::EventData *, HdbEventDataType>> events) override;

	// Inserts the attribute configuration data (Tango Attribute Configuration event data)
	// into the database. The attribute must be configured to be stored in HDB++,
	// otherwise an exception will be thrown.
	void insert_param_event(Tango::AttrConfEventData *param_event, const HdbEventDataType & /* data_type */) override;

	// Add an attribute to the database. Trying to add an attribute that already exists will
	// cause an exception
	void add_attribute(const std::string &fqdn_attr_name, int type, int format, int write_type) override;

	// Update the attribute ttl. The attribute must have been configured to be stored in
	// HDB++, otherwise an exception is raised
	void update_ttl(const std::string &fqdn_attr_name, unsigned int ttl) override;

	// Inserts a history event for the attribute name passed to the function. The attribute
	// must have been configured to be stored in HDB++, otherwise an exception is raised.
	// This function will also insert an additional CRASH history event before the START
	// history event if the given event parameter is DB_START and if the last history event
	// stored was also a START event.
	void insert_history_event(const std::string &fqdn_attr_name, unsigned char event) override;

	// Check what hdbpp features this library supports. This library supports: TTL, BATCH_INSERTS
	bool supported(HdbppFeatures feature) override;

private:
	void prepare_insert_event(vector<Tango::EventData *> data, const HdbEventDataType &ev_data_type);
	template <typename Type> void extract_and_store(const vector<event_data_param> &event_data, int data_type/*DEV_DOUBLE, DEV_STRING, ..*/, Tango::AttrDataFormat data_format/*SCALAR, SPECTRUM, ..*/, int write_type/*READ, READ_WRITE, ..*/, enum_field_types mysql_value_type, bool _is_unsigned);
	//template <string> void extract_and_store(const vector<event_data_param> &event_data, int data_type/*DEV_DOUBLE, DEV_STRING, ..*/, Tango::AttrDataFormat data_format/*SCALAR, SPECTRUM, ..*/, int write_type/*READ, READ_WRITE, ..*/, const string &table_name, enum_field_types mysql_value_type, bool _is_unsigned);
	template <typename Type> void store_scalar(const vector<event_values_param<Type> > &event_values, int data_type/*DEV_DOUBLE, DEV_STRING, ..*/, int write_type/*READ, READ_WRITE, ..*/, const string & table_name, enum_field_types mysql_value_type, bool _is_unsigned);
	template <typename Type> void store_array(const string &attr, const vector<Type> &value_r, const vector<Type> &value_w, int quality/*ATTR_VALID, ATTR_INVALID, ..*/, const string &error_desc, int data_type/*DEV_DOUBLE, DEV_STRING, ..*/, int write_type/*READ, READ_WRITE, ..*/, Tango::AttributeDimension attr_r_dim, Tango::AttributeDimension attr_w_dim, double ev_time, double rcv_time, const string &table_name, enum_field_types mysql_value_type, bool _is_unsigned, bool isNull);
	template <typename Type> void store_arrays(const vector<event_values_param<Type> > &event_values, int data_type/*DEV_DOUBLE, DEV_STRING, ..*/, int write_type/*READ, READ_WRITE, ..*/, const string & table_name, enum_field_types mysql_value_type, bool _is_unsigned);
	template <typename Type> void store_array_json(const vector<event_values_param<Type> > &event_values, int data_type/*DEV_DOUBLE, DEV_STRING, ..*/, int write_type/*READ, READ_WRITE, ..*/, const string & table_name, enum_field_types mysql_value_type, bool _is_unsigned);

	template <typename Type> bool is_nan_or_inf(Type val);
};

class HdbPPMySQLFactory : public DBFactory
{

public:
	virtual AbstractDB* create_db(const string &id, const vector<string> &configuration);

};

} // namespace hdbpp
#endif

